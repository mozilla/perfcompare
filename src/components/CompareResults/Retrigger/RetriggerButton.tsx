import { useState } from 'react';

import RefreshOutlinedIcon from '@mui/icons-material/RefreshOutlined';
import { IconButton } from '@mui/material';

import {
  getTaskclusterCredentials,
  getTaskclusterParams,
  retrigger,
  signInIntoTaskcluster,
} from '../../../logic/taskcluster';
import {
  fetchDecisionTaskIdFromPushId,
  fetchJobInformationFromJobId,
} from '../../../logic/treeherder';
import { Strings } from '../../../resources/Strings';
import { CompareResultsItem } from '../../../types/state';
import { RetriggerConfigModal } from './RetriggerConfigModal';
import { RetriggerSignInModal } from './RetriggerSignInModal';

type Status = 'pending' | 'signin-modal' | 'retrigger-modal';

interface RetriggerButtonProps {
  result: CompareResultsItem;
}

function RetriggerButton({ result }: RetriggerButtonProps) {
  const {
    base_repository_name: baseRepository,
    base_retriggerable_job_ids: baseRetriggerableJobIds,
    new_repository_name: newRepository,
    new_retriggerable_job_ids: newRetriggerableJobIds,
  } = result;

  const [status, setStatus] = useState('pending' as Status);

  const getRetriggerConfig = async (
    repository: string,
    jobId: number,
    times: number,
  ) => {
    const tcParams = getTaskclusterParams();

    const jobInfo = await fetchJobInformationFromJobId(repository, jobId);
    const decisionTaskId = await fetchDecisionTaskIdFromPushId(
      repository,
      jobInfo.push_id,
    );

    const config = {
      rootUrl: tcParams.url,
      jobInfo,
      decisionTaskId,
      times,
    };

    return config;
  };

  const onRetriggerButtonClick = async () => {
    const credentials = getTaskclusterCredentials();
    if (!credentials) {
      setStatus('signin-modal');
      return;
    }

    setStatus('retrigger-modal');
  };

  const onSignIn = async () => {
    await signInIntoTaskcluster();
    setStatus('retrigger-modal');
  };

  const onRetriggerConfirm = async ({
    baseTimes,
    newTimes,
  }: {
    baseTimes: number;
    newTimes: number;
  }) => {
    setStatus('pending');

    const baseRetriggerConfigPromise = getRetriggerConfig(
      baseRepository,
      baseRetriggerableJobIds[0],
      baseTimes,
    );

    const newRetriggerConfigPromise = getRetriggerConfig(
      newRepository,
      newRetriggerableJobIds[0],
      newTimes,
    );

    const [baseRetriggerTaskId, newRetriggerTaskId] = await Promise.all([
      baseRetriggerConfigPromise.then(retrigger),
      newRetriggerConfigPromise.then(retrigger),
    ]);
    console.log('Retrigger taskId for base: ', baseRetriggerTaskId);
    console.log('Retrigger taskId for new: ', newRetriggerTaskId);
  };

  return (
    <>
      <IconButton
        title={Strings.components.revisionRow.title.retriggerJobs}
        color='primary'
        size='small'
        onClick={() => void onRetriggerButtonClick()}
      >
        <RefreshOutlinedIcon />
      </IconButton>
      <RetriggerSignInModal
        open={status === 'signin-modal'}
        onClose={() => setStatus('pending')}
        onSignIn={onSignIn}
      />
      <RetriggerConfigModal
        open={status === 'retrigger-modal'}
        onClose={() => setStatus('pending')}
        onRetriggerClick={onRetriggerConfirm}
      />
    </>
  );
}

export default RetriggerButton;
