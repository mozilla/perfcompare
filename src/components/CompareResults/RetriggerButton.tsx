import RefreshOutlinedIcon from '@mui/icons-material/RefreshOutlined';
import { IconButton } from '@mui/material';

import {
  getTaskclusterCredentials,
  signInIntoTaskcluster,
} from '../../logic/taskcluster';
import { Strings } from '../../resources/Strings';

function RetriggerButton() {
  const onOpenModal = async () => {
    let credentials = getTaskclusterCredentials();
    if (!credentials) {
      await signInIntoTaskcluster();
      credentials = getTaskclusterCredentials();
    }

    console.log('We have an access token!', credentials);
  };

  // TODO implement modal

  return (
    <>
      <IconButton
        title={Strings.components.revisionRow.title.retriggerJobs}
        color='primary'
        size='small'
        onClick={() => void onOpenModal()}
      >
        <RefreshOutlinedIcon />
      </IconButton>
    </>
  );
}

export default RetriggerButton;
