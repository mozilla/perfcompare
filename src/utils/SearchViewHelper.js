import store from '../common/store';
import {
  updateSearchValue,
  updateSearchResults,
  updateRepository,
} from '../reducers/searchSlice';
import {
  fetchRecentRevisions,
  fetchRevisionByID,
  fetchRevisionsByAuthor,
} from '../thunks/searchThunk';

const SearchViewHelper = {
  searchByRevisionOrEmail(repository, search) {
    const emailMatch = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    const longHashMatch = /[a-zA-Z0-9]{40}/;

    if (emailMatch.test(search)) {
      store.dispatch(fetchRevisionsByAuthor({ repository, search }));
    } else if (longHashMatch.test(search)) {
      store.dispatch(fetchRevisionByID({ repository, search }));
    }
  },

  handleChangeDropdown(event) {
    const repository = event.target.innerText;
    const search = store.getState().search.searchValue;

    // Update state with selected repository
    store.dispatch(updateRepository(repository));
    // If repository is selected after search value is input, fetch search results
    if (search !== '') {
      SearchViewHelper.searchByRevisionOrEmail(repository, search);
    }
    // Fetch 10 most recent revisions when repository changes
    store.dispatch(fetchRecentRevisions(repository));
  },

  handleChangeSearch(event) {
    const search = event.target.value;
    const { repository } = store.getState().search;
    store.dispatch(updateSearchValue(search));

    // If search input is cleared, clear results
    if (search === '') store.dispatch(updateSearchResults([]));

    if (repository === '') {
      console.log('Please select a repository');
    } else {
      SearchViewHelper.searchByRevisionOrEmail(repository, search);
    }
  },
};

export default SearchViewHelper;
