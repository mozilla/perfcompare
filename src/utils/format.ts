const dateFormatter = new Intl.DateTimeFormat('en-US', {
  dateStyle: 'medium',
});

export function formatDate(date: Date) {
  return dateFormatter.format(date);
}

export function formatDateRange(date1: Date, date2: Date) {
  return formatDate(date1) + ' - ' + formatDate(date2);
}
