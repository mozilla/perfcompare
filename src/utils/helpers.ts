import {
  frameworkMap,
  devToolsFramework,
  baseDocsURL,
  removedOldTestDevTools,
  nonDocumentedTestsDevTools,
  supportedPerfdocsFrameworks,
} from '../common/constants';
import type { Repository, Changeset } from '../types/state';
import type { Framework, SupportedPerfdocsFramework } from '../types/types';

const truncateHash = (revision: Changeset['revision']) => revision.slice(0, 12);

// return only most recent commit message
// first commit is usually 'try_task_config'
const getLatestCommitMessage = (item: Changeset) => {
  const { repository_id: repositoryId, revisions } = item;
  const isTry = repositoryId === 4;
  const lastUsefulRevision =
    isTry && revisions.length > 1 ? revisions[1] : revisions[0];
  const lastUsefulSummary = lastUsefulRevision.comments.slice(
    0,
    lastUsefulRevision.comments.indexOf('\n'),
  );
  return lastUsefulSummary;
};

// ensure all numbers display two digits
const formatNumber = (number: number): string => {
  if (number < 10) {
    return `0${number}`;
  } else return String(number);
};

const formatDate = (timestamp: number): string => {
  const date = new Date(timestamp * 1000);
  const day = date.getDate();
  const month = date.getMonth() + 1;
  const year = String(date.getFullYear()).slice(2);
  const hours = date.getHours();
  const minutes = date.getMinutes();

  return `${formatNumber(month)}/${formatNumber(day)}/${year} ${formatNumber(
    hours,
  )}:${formatNumber(minutes)}`;
};

const getTreeherderURL = (
  revision: Changeset['revision'],
  repository: Repository['name'],
) =>
  `https://treeherder.mozilla.org/jobs?repo=${repository}&revision=${revision}`;

const createDevtoolsDocsUrl = (
  supportedFramework: string,
  urlReadySuite: string,
  suite: string,
) => {
  let linkSupported = true;

  let devtoolsDocsURL = `${baseDocsURL}/devtools/tests/${supportedFramework}.html#${urlReadySuite}`;
  if (
    suite === removedOldTestDevTools ||
    nonDocumentedTestsDevTools.includes(suite)
  ) {
    devtoolsDocsURL = '';
    linkSupported = false;
  }
  return { devtoolsDocsURL, linkSupported };
};

const getDocsURL = (suite: string, framework_id: Framework['id']) => {
  const framework = frameworkMap[framework_id];
  const supportedFramework =
    supportedPerfdocsFrameworks[framework as SupportedPerfdocsFramework];
  const urlReadySuite = suite.replace(/:|\s|\.|_/g, '-').toLowerCase();

  let docsURL = '';
  let isLinkSupported = true;

  if (framework_id === devToolsFramework.id) {
    const { devtoolsDocsURL, linkSupported } = createDevtoolsDocsUrl(
      supportedFramework,
      urlReadySuite,
      suite,
    );

    isLinkSupported = linkSupported;
    docsURL = devtoolsDocsURL;
  } else if (supportedFramework) {
    docsURL = `${baseDocsURL}/testing/perfdocs/${supportedFramework}.html#${urlReadySuite}`;
  } else {
    isLinkSupported = false;
  }
  return { docsURL, isLinkSupported };
};

// TO DO: Review if this method is still needed
const setConfidenceClassName = (confidenceText: string | null) => {
  return confidenceText || 'unknown-confidence';
};

const swapArrayElements = <T>(
  array: T[],
  index1: number,
  index2: number,
): T[] => {
  if (
    index1 !== index2 &&
    [index1, index2].every((index) => index < array.length && index >= 0)
  ) {
    const newArray = [...array];
    [newArray[index1], newArray[index2]] = [newArray[index2], newArray[index1]];
    return newArray;
  }
  return array;
};

export {
  formatDate,
  getLatestCommitMessage,
  getTreeherderURL,
  setConfidenceClassName,
  swapArrayElements,
  truncateHash,
  getDocsURL,
};
