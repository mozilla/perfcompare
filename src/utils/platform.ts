import type { PlatformShortName } from '../types/state';

export const getPlatformShortName = (
  platformName: string,
): PlatformShortName => {
  if (platformName.toLowerCase().includes('linux')) return 'Linux';
  if (
    platformName.toLowerCase().includes('osx') ||
    platformName.toLowerCase().includes('os x')
  )
    return 'OSX';
  if (platformName.toLowerCase().includes('windows')) return 'Windows';
  if (platformName.toLowerCase().includes('android')) return 'Android';
  return 'Unspecified';
};
