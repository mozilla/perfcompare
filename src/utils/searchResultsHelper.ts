import type { Revision } from '../types/state';

const truncateHash = (revision: Revision['revision']) => revision.slice(0, 12);

// return only most recent commit message
// first commit is usually 'try_task_config'
const getLatestCommitMessage = (item: Revision) =>
  item.revisions[item.revisions.length - 1].comments;

export { truncateHash, getLatestCommitMessage };
