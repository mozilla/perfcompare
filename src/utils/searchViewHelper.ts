import type { Repository, Revision } from '../types/state';

const truncateHash = (revision: Revision['revision']) => revision.slice(0, 12);

// return only most recent commit message
// first commit is usually 'try_task_config'
const getLatestCommitMessage = (item: Revision) =>
  item.revisions[item.revisions.length - 1].comments;

// ensure all numbers display two digits
const formatNumber = (number: number): string => {
  if (number < 10) {
    return `0${number}`;
  } else return String(number);
};

const formatDate = (timestamp: number): string => {
  const date = new Date(timestamp * 1000);
  const day = date.getDate();
  const month = date.getMonth() + 1;
  const year = String(date.getFullYear()).slice(2);
  const hours = date.getHours();
  const minutes = date.getMinutes();

  return `${formatNumber(month)}/${formatNumber(day)}/${year} ${formatNumber(
    hours,
  )}:${formatNumber(minutes)}`;
};

const getTreeherderURL = (
  revision: Revision['revision'],
  repository: Repository['name'],
) =>
  `https://treeherder.mozilla.org/jobs?repo=${repository}&revision=${revision}`;

export { formatDate, getLatestCommitMessage, getTreeherderURL, truncateHash };
