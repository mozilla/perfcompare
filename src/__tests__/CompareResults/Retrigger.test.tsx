import userEvent from '@testing-library/user-event';
import { Hooks } from 'taskcluster-client-web';

import RetriggerButton from '../../components/CompareResults/Retrigger/RetriggerButton';
import { getLocationOrigin } from '../../utils/location';
import getTestData from '../utils/fixtures';
import {
  FetchMockSandbox,
  fireEvent,
  render,
  screen,
  waitFor,
  within,
} from '../utils/test-utils';

jest.mock('../../utils/location');
const mockedGetLocationOrigin = getLocationOrigin as jest.Mock;
const result = getTestData().testCompareData[0];

const retriggerMultipleActionDefinition = {
  context: [],
  description: 'Create a clone of the task.',
  extra: {
    actionPerm: 'generic',
  },
  hookGroupId: 'project-gecko',
  hookId: 'in-tree-action-3-generic/f526da500b',
  hookPayload: {
    decision: {
      action: {
        cb_name: 'retrigger-multiple',
        description: 'Create a clone of the task.',
        name: 'retrigger-multiple',
        symbol: 'rt',
        taskGroupId: 'ZccXq7SFRLyRF7l6unyxkQ',
        title: 'Retrigger',
      },
      push: {
        base_revision: '51748d809eb5e6a7c86eeeb911f3b2b12324b1bf',
        owner: 'mozilla-taskcluster-maintenance@mozilla.com',
        pushlog_id: '42024',
        revision: 'ab17014f4b8f237c9d5fe7b88f8a89c49d0ffe7c',
      },
      repository: {
        level: '3',
        project: 'mozilla-central',
        url: 'https://hg.mozilla.org/mozilla-central',
      },
    },
    user: {
      input: {
        $eval: 'input',
      },
      taskGroupId: {
        $eval: 'taskGroupId',
      },
      taskId: {
        $eval: 'taskId',
      },
    },
  },
  kind: 'hook',
  name: 'retrigger-multiple',
  schema: {
    properties: {
      additionalProperties: false,
      requests: {
        items: {
          additionalProperties: false,
          tasks: {
            description: 'An array of task labels',
            items: {
              type: 'string',
            },
            type: 'array',
          },
          times: {
            description: 'How many times to run each task.',
            maximum: 100,
            minimum: 1,
            title: 'Times',
            type: 'integer',
          },
        },
        type: 'array',
      },
    },
    type: 'object',
  },
  title: 'Retrigger',
};

const setUpUserCredentials = () => {
  sessionStorage.setItem('requestState', 'OkCrH5isZncYqeJbRDelN');
  sessionStorage.setItem(
    'taskclusterUrl',
    'https://firefox-ci-tc.services.mozilla.com',
  );
  window.localStorage.setItem(
    'userTokens',
    JSON.stringify({
      'https://firefox-ci-tc.services.mozilla.com': {
        access_token: 'RnVpOGJtdDZTb3FlWW5PVUxVclprQQ==',
        token_type: 'Bearer',
      },
    }),
  );
  window.localStorage.setItem(
    'userCredentials',
    JSON.stringify({
      'https://firefox-ci-tc.services.mozilla.com': {
        expires: '2044-05-20T14:07:40.828Z',
        credentials: {
          clientId:
            'mozilla-auth0/ad|Mozilla-LDAP|ldapuser/perfcompare-localhost-3000-client-OCvzh5',
          accessToken: 'jQWJVQdeRceT-YymPwTWagPh2PwJr0RmmZyL1uAfMSWg',
        },
      },
    }),
  );
};

describe('Retrigger', () => {
  beforeEach(() => {
    mockedGetLocationOrigin.mockImplementation(() => 'http://localhost:3000');
  });

  it('should display Sign In modal when there are no credentials', async () => {
    render(<RetriggerButton result={result} />);

    const openModalButton = await screen.findByTitle('retrigger jobs');

    const user = userEvent.setup({ delay: null });
    await user.click(openModalButton);

    const signInButton = await screen.findByText('Sign in');

    expect(signInButton).toBeInTheDocument();
  });

  it('should retrigger one job for base revision and one job for new revision', async () => {
    // fetch requests for base revision info
    (window.fetch as FetchMockSandbox)
      .get(
        'begin:https://treeherder.mozilla.org/api/project/mozilla-central/jobs/381594973/',
        {
          push_id: 1234567,
        },
      )
      .get(
        'begin:https://treeherder.mozilla.org/api/project/mozilla-central/push/decisiontask/?push_ids=1234567',
        {
          '1234567': { id: 'TeSt0FIBQyuzPDktQnTest', run: '0' },
        },
      )
      .get(
        'begin:https://firefox-ci-tc.services.mozilla.com/api/queue/v1/task/TeSt0FIBQyuzPDktQnTest/artifacts/public%2Factions.json',
        {
          actions: [{ ...retriggerMultipleActionDefinition }],
          variables: {},
        },
      )
      .get(
        'begin:https://firefoxci.taskcluster-artifacts.net/TeSt0FIBQyuzPDktQnTest/0/public/actions.json',
        {
          actions: [{ ...retriggerMultipleActionDefinition }],
          variables: {},
        },
      );
    // fetch requests for new revision info
    (window.fetch as FetchMockSandbox)
      .get(
        'begin:https://treeherder.mozilla.org/api/project/mozilla-central/jobs/381452501/',
        {
          push_id: 7891011,
        },
      )
      .get(
        'begin:https://treeherder.mozilla.org/api/project/mozilla-central/push/decisiontask/?push_ids=7891011',
        {
          '7891011': { id: 'TeStSeCoNDuzPDktwoTest', run: '0' },
        },
      )
      .get(
        'https://firefox-ci-tc.services.mozilla.com/api/queue/v1/task/TeStSeCoNDuzPDktwoTest/artifacts/public%2Factions.json',
        {
          actions: [{ ...retriggerMultipleActionDefinition }],
          variables: {},
        },
      )
      .get(
        'begin:https://firefoxci.taskcluster-artifacts.net/TeStSeCoNDuzPDktwoTest/0/public/actions.json',
        {
          actions: [{ ...retriggerMultipleActionDefinition }],
          variables: {},
        },
      );

    setUpUserCredentials();
    render(<RetriggerButton result={result} />);

    const openModalButton = await screen.findByRole('button', {
      name: 'retrigger jobs',
    });

    const user = userEvent.setup({ delay: null });
    await user.click(openModalButton);

    // select option 1 for base
    const baseSelect = await screen.findByLabelText('Base');

    fireEvent.mouseDown(baseSelect);

    let listbox = within(await screen.findByRole('listbox'));
    fireEvent.click(await listbox.findByText('1'));

    // select option 1 for new
    const newSelect = await screen.findByLabelText('New');

    fireEvent.mouseDown(newSelect);

    listbox = within(await screen.findByRole('listbox'));
    fireEvent.click(await listbox.findByText('1'));

    const triggerJobsButton = await screen.findByRole('button', {
      name: 'Retrigger',
    });
    const MockedHooks = Hooks as jest.Mock;

    await user.click(triggerJobsButton);

    await waitFor(() =>
      expect(new MockedHooks().triggerHook).toHaveBeenCalled(),
    );
  });
});
