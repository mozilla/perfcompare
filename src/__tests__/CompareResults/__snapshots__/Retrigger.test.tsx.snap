// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Retrigger should retrigger one job for base revision and one job for new revision 1`] = `
<div
  aria-describedby="notistack-snackbar"
  class="SnackbarContent-root SnackbarItem-contentRoot SnackbarItem-variantInfo SnackbarItem-lessPadding css-hped4j"
  role="alert"
  style="webkit-transform: none; transform: none; webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
>
  <div
    class="SnackbarItem-message"
    id="notistack-snackbar"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
      focusable="false"
      style="font-size: 20px; margin-inline-end: 8px;"
      viewBox="0 0 24 24"
    >
      <path
        d="M13,9H11V7H13M13,17H11V11H13M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,
        0 22,12A10,10 0 0,0 12,2Z"
      />
    </svg>
    The retrigger request for the base run has been sent successfully, with ID rEtrRigGERtaSkId.
  </div>
  <div
    class="SnackbarItem-action"
  >
    <a
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation css-aa1ild-MuiButtonBase-root-MuiButton-root"
      href="https://treeherder.mozilla.org/jobs?repo=mozilla-central&revision=coconut"
      rel="noreferrer"
      tabindex="0"
      target="_blank"
    >
      Open Treeherder
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </a>
    <button
      aria-label="Dismiss"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-16podhg-MuiButtonBase-root-MuiIconButton-root"
      data-testid="alert-close"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
        data-testid="CloseIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`Retrigger should retrigger one job for base revision and one job for new revision 2`] = `
<div
  aria-describedby="notistack-snackbar"
  class="SnackbarContent-root SnackbarItem-contentRoot SnackbarItem-variantInfo SnackbarItem-lessPadding css-hped4j"
  role="alert"
  style="webkit-transform: none; transform: none; webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
>
  <div
    class="SnackbarItem-message"
    id="notistack-snackbar"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
      focusable="false"
      style="font-size: 20px; margin-inline-end: 8px;"
      viewBox="0 0 24 24"
    >
      <path
        d="M13,9H11V7H13M13,17H11V11H13M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,
        0 22,12A10,10 0 0,0 12,2Z"
      />
    </svg>
    The retrigger request for the new run has been sent successfully, with ID rEtrRigGERtaSkId.
  </div>
  <div
    class="SnackbarItem-action"
  >
    <a
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation css-aa1ild-MuiButtonBase-root-MuiButton-root"
      href="https://treeherder.mozilla.org/jobs?repo=mozilla-central&revision=spam"
      rel="noreferrer"
      tabindex="0"
      target="_blank"
    >
      Open Treeherder
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </a>
    <button
      aria-label="Dismiss"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-16podhg-MuiButtonBase-root-MuiIconButton-root"
      data-testid="alert-close"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
        data-testid="CloseIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
